File: src/schur-weyl/sw_measure.py
import numpy as np
from young_diagrams import partitions
from dimensions import dim_specht
from symmetric_functions import schur_polynomial


def schur_weyl_measure(spectrum: list[float], k: int) -> dict[tuple[int, ...], float]:
    """The probabilities $tr(\\Pi^\\lambda rho^{\\otimes k})$ where rho has the defined spectrum

    Args:
        spectrum (list[float]): the spectrum of rho
        k (int): the number of copies of rho

    Returns:
        dict[tuple[int, ...], float]: The probabilities and their associated partition.
    """
    ret: dict[tuple[int, ...], float] = dict()
    for lam in partitions(k):
        ret[lam] = dim_specht(lam) * schur_polynomial(lam, spectrum)
    return ret

File: docs/src/schur-weyl/swMeasure.ts
/**
 * Ported from `src/schur_weyl/sw_measure.py`.
 */

import type { Partition } from './youngDiagrams';
import { partitions } from './youngDiagrams';
import { dimSpecht } from './dimensions';
import { schurPolynomial } from './symmetricFunctions';

export interface SchurWeylEntry {
  partition: Partition;
  probability: number;
}

/**
 * The probabilities tr(Pi^lam rho^{\otimes k}) where rho has the given
 * spectrum, one entry per partition of k.
 *
 * @param spectrum - the spectrum of rho
 * @param k - the number of copies of rho
 */
export function schurWeylMeasure(spectrum: number[], k: number): SchurWeylEntry[] {
  const ret: SchurWeylEntry[] = [];
  for (const lam of partitions(k)) {
    ret.push({ partition: lam, probability: dimSpecht(lam) * schurPolynomial(lam, spectrum) });
  }
  return ret;
}
